src/framework/MOM_array_transform.F90

```diff
@@ -0,0 +1,358 @@
+subroutine rotate_array_real_4d(A_in, turns, A)
+  real, intent(in) :: A_in(:,:,:,:) !< Unrotated array
+  integer, intent(in) :: turns      !< Number of quarter turns
+  real, intent(out) :: A(:,:,:,:)   !< Rotated array
+
+  integer :: n
+
+  do n = lbound(A_in, 4), ubound(A_in, 4)
+    call rotate_array(A_in(:,:,:,n), turns, A(:,:,:,n))
+  enddo
+end subroutine rotate_array_real_4d
+
+
+!> Rotate the elements of a 2d integer array along first and second axes.
+subroutine rotate_array_integer(A_in, turns, A)
+  integer, intent(in) :: A_in(:,:)  !< Unrotated array
+  integer, intent(in) :: turns      !< Number of quarter turns
+  integer, intent(out) :: A(:,:)    !< Rotated array
+
+  integer :: m, n
+
+  m = size(A_in, 1)
+  n = size(A_in, 2)
+
+  select case (modulo(turns, 4))
+    case(0)
+      A = A_in
+    case(1)
+      A = transpose(A_in)
+      A = A(n:1:-1, :)
+    case(2)
+      A = A_in(m:1:-1, n:1:-1)
+    case(3)
+      A = transpose(A_in(m:1:-1, :))
+  end select
+end subroutine rotate_array_integer
+
+
+!> Rotate the elements of a 2d logical array along first and second axes.
+subroutine rotate_array_logical(A_in, turns, A)
+  logical, intent(in) :: A_in(:,:)  !< Unrotated array
+  integer, intent(in) :: turns      !< Number of quarter turns
+  logical, intent(out) :: A(:,:)    !< Rotated array
+
+  integer :: m, n
+
+  m = size(A_in, 1)
+  n = size(A_in, 2)
+
+  select case (modulo(turns, 4))
+    case(0)
+      A = A_in
+    case(1)
+      A = transpose(A_in)
+      A = A(n:1:-1, :)
+    case(2)
+      A = A_in(m:1:-1, n:1:-1)
+    case(3)
+      A = transpose(A_in(m:1:-1, :))
+  end select
+end subroutine rotate_array_logical
+
+
+!> Rotate the elements of a 2d real array pair along first and second axes.
+subroutine rotate_array_pair_real_2d(A_in, B_in, turns, A, B)
+  real, intent(in) :: A_in(:,:)   !< Unrotated scalar array pair
+  real, intent(in) :: B_in(:,:)   !< Unrotated scalar array pair
+  integer, intent(in) :: turns    !< Number of quarter turns
+  real, intent(out) :: A(:,:)     !< Rotated scalar array pair
+  real, intent(out) :: B(:,:)     !< Rotated scalar array pair
+
+  if (modulo(turns, 2) /= 0) then
+    call rotate_array(B_in, turns, A)
+    call rotate_array(A_in, turns, B)
+  else
+    call rotate_array(A_in, turns, A)
+    call rotate_array(B_in, turns, B)
+  endif
+end subroutine rotate_array_pair_real_2d
+
+
+!> Rotate the elements of a 3d real array pair along first and second axes.
+subroutine rotate_array_pair_real_3d(A_in, B_in, turns, A, B)
+  real, intent(in) :: A_in(:,:,:)   !< Unrotated scalar array pair
+  real, intent(in) :: B_in(:,:,:)   !< Unrotated scalar array pair
+  integer, intent(in) :: turns      !< Number of quarter turns
+  real, intent(out) :: A(:,:,:)     !< Rotated scalar array pair
+  real, intent(out) :: B(:,:,:)     !< Rotated scalar array pair
+
+  integer :: k
+
+  do k = lbound(A_in, 3), ubound(A_in, 3)
```

Comment:
Since these dummy arrays are 1-indexed by default, use of `lbound` and `ubound` is not adding any value and potentially causing confusion, so I will rewrite this and the other loops as 1-to-n.